File: src/protocols/swarm-intelligence-protocol.js
/**
 * PROTO-233: Swarm Intelligence Protocol
 * ============================================================
 * Particle Swarm Optimization, Ant Colony pheromone trails, and
 * Swarm Consensus for Modular Memory Intelligence Multi-Systems (MMIMS-X).
 *
 * Implements:
 *  - PSO with phi-weighted inertia / cognitive / social coefficients
 *  - Pheromone trail system for ACO-style routing
 *  - Swarm consensus over discrete options
 *  - Integration of cognitive, social, and inertia vectors
 *
 * Charter: PROTO-233
 * Attribution: Alfredo Medina Hernandez | Medina Tech | Dallas TX | May 2026
 */

'use strict';

// ─── Constants ────────────────────────────────────────────────────────────────
const PHI       = (1 + Math.sqrt(5)) / 2;   // 1.618033…
const PHI_INV   = 1 / PHI;                  // 0.618033…
const PHI_SQ    = PHI * PHI;                // 2.618033…

/**
 * Phi-weighted PSO coefficients.
 *   w  (inertia)  = φ⁻²  ≈ 0.382
 *   c₁ (cognitive)= φ⁻¹  ≈ 0.618
 *   c₂ (social)   = φ     ≈ 1.618
 *
 * Sum w + c₁ + c₂ ≈ 2.618 = φ² — a meaningful phi relationship.
 */
const PSO_W  = 1 / PHI_SQ;   // inertia weight ≈ 0.382
const PSO_C1 = PHI_INV;      // cognitive coefficient ≈ 0.618
const PSO_C2 = PHI;           // social coefficient ≈ 1.618

// Pheromone evaporation rate (phi-inverse per step)
const EVAPORATION = PHI_INV * 0.1;   // ≈ 0.0618 per step

// ─── Utilities ────────────────────────────────────────────────────────────────
function randUniform()           { return Math.random(); }
function clamp(v, lo, hi)        { return Math.max(lo, Math.min(hi, v)); }
function vecAdd(a, b)            { return a.map((x, i) => x + b[i]); }
function vecSub(a, b)            { return a.map((x, i) => x - b[i]); }
function vecScale(a, s)          { return a.map(x => x * s); }
function vecDot(a, b)            { return a.reduce((s, x, i) => s + x * b[i], 0); }
function randVec(n)              { return Array.from({ length: n }, randUniform); }
function vecClamp(v, lo, hi)     { return v.map(x => clamp(x, lo, hi)); }

// ─── Particle (PSO) ───────────────────────────────────────────────────────────
class Particle {
  /**
   * @param {number[]} position — initial position in search space
   * @param {number[]} velocity — initial velocity
   * @param {Function} fitnessFn — (position) => number (higher = better)
   */
  constructor(position, velocity, fitnessFn) {
    this.position   = [...position];
    this.velocity   = [...velocity];
    this.fitnessFn  = fitnessFn;
    this.bestPos    = [...position];
    this.bestFit    = fitnessFn(position);
    this.id         = `P-${Date.now()}-${Math.random().toString(36).slice(2, 6)}`;
    this.dim        = position.length;
  }

  /**
   * Update velocity and position with phi-weighted PSO equations.
   *   v_i(t+1) = w·v_i(t) + c₁·r₁·(pBest - x) + c₂·r₂·(gBest - x)
   *   x_i(t+1) = x_i(t) + v_i(t+1)
   */
  step(globalBest, bounds) {
    const r1  = randUniform();
    const r2  = randUniform();
    const cog = vecScale(vecSub(this.bestPos,   this.position), PSO_C1 * r1);
    const soc = vecScale(vecSub(globalBest,      this.position), PSO_C2 * r2);
    const ine = vecScale(this.velocity, PSO_W);

    this.velocity = vecAdd(vecAdd(ine, cog), soc);

    // Bound velocity to [±range×φ] per dimension
    if (bounds) {
      this.velocity = this.velocity.map((v, i) => {
        const span = (bounds[i][1] - bounds[i][0]) * PHI_INV;
        return clamp(v, -span, span);
      });
    }

    this.position = vecAdd(this.position, this.velocity);
    if (bounds) {
      this.position = this.position.map((x, i) => clamp(x, bounds[i][0], bounds[i][1]));
    }

    const fit = this.fitnessFn(this.position);
    if (fit > this.bestFit) {
      this.bestFit = fit;
      this.bestPos = [...this.position];
    }
    return fit;
  }
}

// ─── Particle Swarm Optimizer ─────────────────────────────────────────────────
class ParticleSwarmOptimizer {
  /**
   * @param {Function}   fitnessFn  — (pos: number[]) => number
   * @param {number}     dim        — dimensionality
   * @param {number[][]} [bounds]   — [[lo,hi], ...] per dimension
   * @param {number}     [nParticles] — default 34 (Fibonacci)
   */
  constructor(fitnessFn, dim, bounds = null, nParticles = 34) {
    this.fitnessFn  = fitnessFn;
    this.dim        = dim;
    this.bounds     = bounds;
    this.nParticles = nParticles;
    this.particles  = [];
    this.globalBest = null;
    this.globalFit  = -Infinity;
    this.iteration  = 0;
    this.history    = [];

    this._init();
  }

  _init() {
    for (let i = 0; i < this.nParticles; i++) {
      const pos = this.bounds
        ? this.bounds.map(([lo, hi]) => lo + randUniform() * (hi - lo))
        : randVec(this.dim);
      const vel = this.bounds
        ? this.bounds.map(([lo, hi]) => (randUniform() - 0.5) * (hi - lo) * PHI_INV)
        : Array.from({ length: this.dim }, () => (randUniform() - 0.5) * PHI_INV);
      const p = new Particle(pos, vel, this.fitnessFn);
      this.particles.push(p);
      if (p.bestFit > this.globalFit) {
        this.globalFit  = p.bestFit;
        this.globalBest = [...p.bestPos];
      }
    }
  }

  /** Run one iteration */
  step() {
    this.iteration++;
    for (const p of this.particles) {
      const fit = p.step(this.globalBest, this.bounds);
      if (fit > this.globalFit) {
        this.globalFit  = fit;
        this.globalBest = [...p.position];
      }
    }
    this.history.push({ iter: this.iteration, fitness: this.globalFit });
    return { iteration: this.iteration, best: this.globalBest, fitness: this.globalFit };
  }

  /** Run n iterations */
  run(n) {
    let last;
    for (let i = 0; i < n; i++) last = this.step();
    return last;
  }
}

// ─── Pheromone Trail System (ACO) ─────────────────────────────────────────────
class PheromoneTrailSystem {
  /**
   * Directed graph of pheromone intensities between nodes.
   * @param {number} [initialLevel] — initial pheromone on all trails
   */
  constructor(initialLevel = PHI_INV) {
    this.trails     = new Map();  // `${from}:${to}` → level
    this.initial    = initialLevel;
    this.stepCount  = 0;
  }

  _key(from, to) { return `${from}:${to}`; }

  getLevel(from, to) {
    return this.trails.get(this._key(from, to)) ?? this.initial;
  }

  /** Deposit pheromone proportional to solution quality Q */
  deposit(path, quality = 1.0) {
    const delta = (quality * PHI) / (path.length || 1);
    for (let i = 0; i < path.length - 1; i++) {
      const key = this._key(path[i], path[i + 1]);
      this.trails.set(key, (this.trails.get(key) ?? this.initial) + delta);
    }
  }

  /** Evaporate all trails by phi-inverse factor */
  evaporate() {
    this.stepCount++;
    for (const [key, level] of this.trails) {
      const next = level * (1 - EVAPORATION);
      if (next < 1e-6) this.trails.delete(key);
      else             this.trails.set(key, next);
    }
  }

  /**
   * Probabilistically choose next node from current, given candidate list.
   * P(j) ∝ pheromone(i→j)^α × heuristic(j)^β   (α=φ, β=φ⁻¹)
   */
  chooseNext(from, candidates, heuristicFn) {
    const alpha = PHI;
    const beta  = PHI_INV;
    const weights = candidates.map(c => {
      const ph = Math.pow(this.getLevel(from, c), alpha);
      const he = Math.pow(heuristicFn(from, c), beta);
      return ph * he;
    });
    const total = weights.reduce((s, w) => s + w, 0);
    let r = randUniform() * total;
    for (let i = 0; i < candidates.length; i++) {
      r -= weights[i];
      if (r <= 0) return candidates[i];
    }
    return candidates[candidates.length - 1];
  }

  stats() {
    const levels = [...this.trails.values()];
    return {
      trails:    this.trails.size,
      evapSteps: this.stepCount,
      maxLevel:  levels.length ? Math.max(...levels) : 0,
      minLevel:  levels.length ? Math.min(...levels) : 0,
      avgLevel:  levels.length ? levels.reduce((s,v) => s+v, 0) / levels.length : 0,
    };
  }
}

// ─── Swarm Consensus ──────────────────────────────────────────────────────────
/**
 * Collective decision-making over a discrete option set.
 * Each agent votes; votes are phi-weighted by agent confidence.
 */
class SwarmConsensus {
  /**
   * @param {string[]} options — discrete options to vote on
   */
  constructor(options) {
    this.options = options;
    this.votes   = new Map();  // agentId → { option, confidence }
    this.round   = 0;
  }

  /**
   * Cast / update vote.
   * @param {string} agentId
   * @param {string} option     — must be in this.options
   * @param {number} confidence — [0, 1]; phi-weighted internally
   */
  vote(agentId, option, confidence = PHI_INV) {
    if (!this.options.includes(option)) throw new Error(`Unknown option: ${option}`);
    this.votes.set(agentId, { option, confidence: clamp(confidence, 0, 1) });
  }

  /**
   * Tally votes with phi-weighting.
   * Returns consensus option + certainty score in [0,1].
   */
  tally() {
    this.round++;
    const tally = {};
    this.options.forEach(o => (tally[o] = 0));

    for (const { option, confidence } of this.votes.values()) {
      tally[option] += confidence * PHI;  // phi-amplified confidence
    }

    const total = Object.values(tally).reduce((s, v) => s + v, 0) || 1;
    const ranked = Object.entries(tally)
      .map(([opt, score]) => ({ option: opt, score, proportion: score / total }))
      .sort((a, b) => b.score - a.score);

    const winner = ranked[0];
    // Certainty: how much the winner dominates (proportional to phi ratio with runner-up)
    const runnerUp = ranked[1]?.score ?? 0;
    const certainty = runnerUp > 0 ? Math.min((winner.score / runnerUp) * PHI_INV, 1) : 1;

    return {
      consensus:  winner.option,
      certainty,
      round:      this.round,
      agentCount: this.votes.size,
      ranking:    ranked,
    };
  }

  /** Reset all votes (start new round) */
  reset() {
    this.votes.clear();
  }
}

// ─── MMIMS-X Swarm Intelligence Bus ───────────────────────────────────────────
/**
 * Unified swarm intelligence layer for MMIMS-X.
 * Integrates PSO optimizers, pheromone trails, and consensus modules.
 */
class SwarmIntelligenceBus {
  constructor() {
    this.optimizers  = new Map();   // id → ParticleSwarmOptimizer
    this.pheromones  = new Map();   // id → PheromoneTrailSystem
    this.consensuses = new Map();   // id → SwarmConsensus
    this.beat        = 0;
  }

  // PSO
  registerOptimizer(id, fitnessFn, dim, bounds = null, nParticles = 34) {
    const pso = new ParticleSwarmOptimizer(fitnessFn, dim, bounds, nParticles);
    this.optimizers.set(id, pso);
    return pso;
  }
  optimize(id, iterations = 55) {
    const pso = this.optimizers.get(id);
    return pso ? pso.run(iterations) : null;
  }

  // ACO
  registerPheromoneTrail(id, initialLevel = PHI_INV) {
    const trail = new PheromoneTrailSystem(initialLevel);
    this.pheromones.set(id, trail);
    return trail;
  }
  getPheromone(id) { return this.pheromones.get(id) ?? null; }

  // Consensus
  registerConsensus(id, options) {
    const c = new SwarmConsensus(options);
    this.consensuses.set(id, c);
    return c;
  }
  vote(id, agentId, option, confidence)  {
    this.consensuses.get(id)?.vote(agentId, option, confidence);
  }
  tally(id) { return this.consensuses.get(id)?.tally() ?? null; }

  /** Global heartbeat: evaporate all pheromone trails */
  tick() {
    this.beat++;
    for (const trail of this.pheromones.values()) trail.evaporate();
    return { beat: this.beat, phiPulse: (this.beat * PHI_INV) % 1 };
  }

  report() {
    return {
      beat:      this.beat,
      optimizers: this.optimizers.size,
      pheromones: this.pheromones.size,
      consensuses: this.consensuses.size,
    };
  }
}

// ─── Exports ──────────────────────────────────────────────────────────────────
module.exports = {
  PHI,
  PHI_INV,
  PHI_SQ,
  PSO_W,
  PSO_C1,
  PSO_C2,
  EVAPORATION,
  Particle,
  ParticleSwarmOptimizer,
  PheromoneTrailSystem,
  SwarmConsensus,
  SwarmIntelligenceBus,
};

File: src/protocols/python/swarm_intelligence_protocol.py
"""
PROTO-233 — Swarm Intelligence Protocol (Python)
PSO, ACO pheromone trails, and swarm consensus for MMIMS-X.

Charter: PROTO-233
Attribution: Alfredo Medina Hernandez | Medina Tech | Dallas TX | May 2026
"""

from __future__ import annotations
import math
import random
import time
from typing import Any, Callable, Dict, List, Optional, Tuple

# ── Constants ──────────────────────────────────────────────────────────────────
PHI         = (1 + math.sqrt(5)) / 2
PHI_INV     = 1 / PHI
PHI_SQ      = PHI * PHI
PSO_W       = 1 / PHI_SQ    # inertia weight   ≈ 0.382
PSO_C1      = PHI_INV       # cognitive coeff  ≈ 0.618
PSO_C2      = PHI           # social coeff     ≈ 1.618
EVAPORATION = PHI_INV * 0.1  # pheromone evap  ≈ 0.0618


# ── Vector Utilities ───────────────────────────────────────────────────────────
def vec_add(a: List[float], b: List[float]) -> List[float]:
    return [x + y for x, y in zip(a, b)]

def vec_sub(a: List[float], b: List[float]) -> List[float]:
    return [x - y for x, y in zip(a, b)]

def vec_scale(a: List[float], s: float) -> List[float]:
    return [x * s for x in a]

def clamp(v: float, lo: float, hi: float) -> float:
    return max(lo, min(hi, v))

def rand_vec(n: int) -> List[float]:
    return [random.random() for _ in range(n)]


# ── Particle (PSO) ─────────────────────────────────────────────────────────────
class Particle:
    def __init__(self, position: List[float], velocity: List[float],
                 fitness_fn: Callable[[List[float]], float],
                 bounds: Optional[List[Tuple[float, float]]] = None):
        self.position   = list(position)
        self.velocity   = list(velocity)
        self.fitness_fn = fitness_fn
        self.bounds     = bounds
        self.best_pos   = list(position)
        self.best_fit   = fitness_fn(position)
        self.id         = f"P-{int(time.time()*1000)}-{random.randint(0,9999)}"
        self.dim        = len(position)

    def step(self, global_best: List[float]) -> float:
        r1, r2 = random.random(), random.random()
        ine = vec_scale(self.velocity, PSO_W)
        cog = vec_scale(vec_sub(self.best_pos, self.position), PSO_C1 * r1)
        soc = vec_scale(vec_sub(global_best,  self.position), PSO_C2 * r2)

        self.velocity = vec_add(vec_add(ine, cog), soc)

        if self.bounds:
            self.velocity = [
                clamp(v, -(b[1]-b[0])*PHI_INV, (b[1]-b[0])*PHI_INV)
                for v, b in zip(self.velocity, self.bounds)
            ]

        self.position = vec_add(self.position, self.velocity)

        if self.bounds:
            self.position = [
                clamp(x, b[0], b[1])
                for x, b in zip(self.position, self.bounds)
            ]

        fit = self.fitness_fn(self.position)
        if fit > self.best_fit:
            self.best_fit = fit
            self.best_pos = list(self.position)
        return fit


# ── Particle Swarm Optimizer ───────────────────────────────────────────────────
class ParticleSwarmOptimizer:
    def __init__(self, fitness_fn: Callable[[List[float]], float],
                 dim: int,
                 bounds: Optional[List[Tuple[float, float]]] = None,
                 n_particles: int = 34):
        self.fitness_fn  = fitness_fn
        self.dim         = dim
        self.bounds      = bounds
        self.n_particles = n_particles
        self.particles:  List[Particle] = []
        self.global_best: List[float] = []
        self.global_fit: float = float("-inf")
        self.iteration  = 0
        self.history:   List[Dict[str, Any]] = []
        self._init()

    def _init(self) -> None:
        for _ in range(self.n_particles):
            if self.bounds:
                pos = [b[0] + random.random() * (b[1]-b[0]) for b in self.bounds]
                vel = [(random.random()-0.5) * (b[1]-b[0]) * PHI_INV for b in self.bounds]
            else:
                pos = rand_vec(self.dim)
                vel = [(random.random()-0.5) * PHI_INV for _ in range(self.dim)]
            p = Particle(pos, vel, self.fitness_fn, self.bounds)
            self.particles.append(p)
            if p.best_fit > self.global_fit:
                self.global_fit  = p.best_fit
                self.global_best = list(p.best_pos)

    def step(self) -> Dict[str, Any]:
        self.iteration += 1
        for p in self.particles:
            fit = p.step(self.global_best)
            if fit > self.global_fit:
                self.global_fit  = fit
                self.global_best = list(p.position)
        result = {"iteration": self.iteration,
                  "best": list(self.global_best),
                  "fitness": self.global_fit}
        self.history.append(result)
        return result

    def run(self, n: int) -> Dict[str, Any]:
        r = None
        for _ in range(n):
            r = self.step()
        return r  # type: ignore


# ── Pheromone Trail System (ACO) ───────────────────────────────────────────────
class PheromoneTrailSystem:
    def __init__(self, initial_level: float = None):
        self.initial   = initial_level if initial_level is not None else PHI_INV
        self._trails:  Dict[str, float] = {}
        self.step_count = 0

    def _key(self, frm: str, to: str) -> str:
        return f"{frm}:{to}"

    def get_level(self, frm: str, to: str) -> float:
        return self._trails.get(self._key(frm, to), self.initial)

    def deposit(self, path: List[str], quality: float = 1.0) -> None:
        delta = (quality * PHI) / max(len(path) - 1, 1)
        for i in range(len(path) - 1):
            k = self._key(path[i], path[i+1])
            self._trails[k] = self._trails.get(k, self.initial) + delta

    def evaporate(self) -> None:
        self.step_count += 1
        dead = []
        for k, v in self._trails.items():
            nv = v * (1 - EVAPORATION)
            if nv < 1e-6:
                dead.append(k)
            else:
                self._trails[k] = nv
        for k in dead:
            del self._trails[k]

    def choose_next(self, frm: str, candidates: List[str],
                    heuristic_fn: Callable[[str, str], float]) -> str:
        alpha, beta = PHI, PHI_INV
        weights = [
            math.pow(self.get_level(frm, c), alpha) * math.pow(heuristic_fn(frm, c), beta)
            for c in candidates
        ]
        total = sum(weights) or 1.0
        r = random.random() * total
        for c, w in zip(candidates, weights):
            r -= w
            if r <= 0:
                return c
        return candidates[-1]

    def stats(self) -> Dict[str, Any]:
        levels = list(self._trails.values())
        return {
            "trails":     len(levels),
            "evap_steps": self.step_count,
            "max_level":  max(levels) if levels else 0,
            "min_level":  min(levels) if levels else 0,
            "avg_level":  sum(levels)/len(levels) if levels else 0,
        }


# ── Swarm Consensus ────────────────────────────────────────────────────────────
class SwarmConsensus:
    def __init__(self, options: List[str]):
        self.options = list(options)
        self._votes:  Dict[str, Tuple[str, float]] = {}  # agentId → (option, confidence)
        self.round    = 0

    def vote(self, agent_id: str, option: str, confidence: float = None) -> None:
        if confidence is None:
            confidence = PHI_INV
        if option not in self.options:
            raise ValueError(f"Unknown option: {option}")
        self._votes[agent_id] = (option, clamp(confidence, 0, 1))

    def tally(self) -> Dict[str, Any]:
        self.round += 1
        scores: Dict[str, float] = {o: 0.0 for o in self.options}
        for opt, conf in self._votes.values():
            scores[opt] += conf * PHI

        total = sum(scores.values()) or 1.0
        ranked = sorted(
            [{"option": o, "score": s, "proportion": s/total} for o, s in scores.items()],
            key=lambda x: x["score"], reverse=True,
        )
        winner    = ranked[0]["score"]
        runner_up = ranked[1]["score"] if len(ranked) > 1 else 0.0
        certainty = min((winner / runner_up) * PHI_INV, 1.0) if runner_up > 0 else 1.0

        return {
            "consensus":   ranked[0]["option"],
            "certainty":   certainty,
            "round":       self.round,
            "agent_count": len(self._votes),
            "ranking":     ranked,
        }

    def reset(self) -> None:
        self._votes.clear()


# ── MMIMS-X Swarm Intelligence Bus ────────────────────────────────────────────
class SwarmIntelligenceBus:
    def __init__(self):
        self.optimizers:  Dict[str, ParticleSwarmOptimizer] = {}
        self.pheromones:  Dict[str, PheromoneTrailSystem]   = {}
        self.consensuses: Dict[str, SwarmConsensus]         = {}
        self.beat         = 0

    def register_optimizer(self, id: str, fitness_fn, dim: int,
                            bounds=None, n_particles: int = 34) -> ParticleSwarmOptimizer:
        pso = ParticleSwarmOptimizer(fitness_fn, dim, bounds, n_particles)
        self.optimizers[id] = pso
        return pso

    def optimize(self, id: str, iterations: int = 55) -> Optional[Dict]:
        pso = self.optimizers.get(id)
        return pso.run(iterations) if pso else None

    def register_pheromone_trail(self, id: str, initial=None) -> PheromoneTrailSystem:
        trail = PheromoneTrailSystem(initial)
        self.pheromones[id] = trail
        return trail

    def get_pheromone(self, id: str) -> Optional[PheromoneTrailSystem]:
        return self.pheromones.get(id)

    def register_consensus(self, id: str, options: List[str]) -> SwarmConsensus:
        c = SwarmConsensus(options)
        self.consensuses[id] = c
        return c

    def vote(self, id: str, agent_id: str, option: str, confidence=None) -> None:
        self.consensuses[id].vote(agent_id, option, confidence)

    def tally(self, id: str) -> Optional[Dict]:
        c = self.consensuses.get(id)
        return c.tally() if c else None

    def tick(self) -> Dict[str, Any]:
        self.beat += 1
        for trail in self.pheromones.values():
            trail.evaporate()
        return {"beat": self.beat, "phi_pulse": (self.beat * PHI_INV) % 1}

    def report(self) -> Dict[str, Any]:
        return {
            "beat":        self.beat,
            "optimizers":  len(self.optimizers),
            "pheromones":  len(self.pheromones),
            "consensuses": len(self.consensuses),
        }


# ── Quick demo ─────────────────────────────────────────────────────────────────
if __name__ == "__main__":
    print("=== PROTO-233 Swarm Intelligence Protocol (Python) ===")

    # PSO: maximise -(x²+y²) → optimum at (0,0)
    pso = ParticleSwarmOptimizer(
        fitness_fn=lambda pos: -(pos[0]**2 + pos[1]**2),
        dim=2,
        bounds=[(-5, 5), (-5, 5)],
        n_particles=34,
    )
    r = pso.run(55)
    print(f"PSO optimum: {r}")

    # ACO
    aco = PheromoneTrailSystem()
    aco.deposit(["A", "B", "C", "D"], quality=1.0)
    print(f"Pheromone A→B: {aco.get_level('A','B'):.4f}")
    aco.evaporate()
    print(f"After evap  A→B: {aco.get_level('A','B'):.4f}")

    # Consensus
    consensus = SwarmConsensus(["route-alpha", "route-beta", "route-gamma"])
    consensus.vote("agent-1", "route-alpha", 0.9)
    consensus.vote("agent-2", "route-alpha", 0.7)
    consensus.vote("agent-3", "route-beta",  0.5)
    print(f"Consensus: {consensus.tally()}")
